File: app-to-app-java/app-to-app/app/src/main/java/com/vonage/tutorial/apptoapp/MainActivity.java
package com.vonage.tutorial.apptoapp;

import android.Manifest;
import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import com.nexmo.client.NexmoCall;
import com.nexmo.client.NexmoCallEventListener;
import com.nexmo.client.NexmoCallHandler;
import com.nexmo.client.NexmoMember;
import com.nexmo.client.NexmoCallMemberStatus;
import com.nexmo.client.NexmoClient;
import com.nexmo.client.NexmoMediaActionState;
import com.nexmo.client.request_listener.NexmoApiError;
import com.nexmo.client.request_listener.NexmoConnectionListener;
import com.nexmo.client.request_listener.NexmoRequestListener;

public class MainActivity extends AppCompatActivity {

    private NexmoClient client;
    private String otherUser = "";
    private NexmoCall onGoingCall;

    private TextView connectionStatusTextView;
    private TextView waitingForIncomingCallTextView;
    private Button loginAsAlice;
    private Button loginAsBob;
    private Button startCallButton;
    private Button answerCallButton;
    private Button rejectCallButton;
    private Button endCallButton;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // init views
        connectionStatusTextView = findViewById(R.id.connectionStatusTextView);
        waitingForIncomingCallTextView = findViewById(R.id.waitingForIncomingCallTextView);
        loginAsAlice = findViewById(R.id.loginAsAlice);
        loginAsBob = findViewById(R.id.loginAsBob);
        startCallButton = findViewById(R.id.startCallButton);
        answerCallButton = findViewById(R.id.answerCallButton);
        rejectCallButton = findViewById(R.id.rejectCallButton);
        endCallButton = findViewById(R.id.endCallButton);

        loginAsAlice.setOnClickListener(v -> loginAsAlice());
        loginAsBob.setOnClickListener(v -> loginAsBob());
        answerCallButton.setOnClickListener(view -> answerCall());
        rejectCallButton.setOnClickListener(view -> rejectCall());
        endCallButton.setOnClickListener(view -> endCall());
        startCallButton.setOnClickListener(v -> startCall());

        // request permissions
        String[] callsPermissions = {Manifest.permission.RECORD_AUDIO};
        ActivityCompat.requestPermissions(this, callsPermissions, 123);

        client = new NexmoClient.Builder().build(this);

        client.setConnectionListener((connectionStatus, connectionStatusReason) -> {
            runOnUiThread(() -> connectionStatusTextView.setText(connectionStatus.toString()));

            if (connectionStatus == NexmoConnectionListener.ConnectionStatus.CONNECTED) {
                runOnUiThread(() -> {
                    hideUI();
                    connectionStatusTextView.setVisibility(View.VISIBLE);
                    startCallButton.setVisibility(View.VISIBLE);
                    waitingForIncomingCallTextView.setVisibility(View.VISIBLE);
                });
            }
        });

        client.addIncomingCallListener(it -> {
            onGoingCall = it;

            runOnUiThread(() -> {
                hideUI();
                answerCallButton.setVisibility(View.VISIBLE);
                rejectCallButton.setVisibility(View.VISIBLE);
            });
        });

    }

    private void hideUI() {
        LinearLayout content = findViewById(R.id.content);

        for (int i = 0; i < content.getChildCount(); i++) {
            View view = content.getChildAt(i);
            view.setVisibility(View.GONE);
        }
    }

    private void loginAsAlice() {
        otherUser = "Bob";

        client.login("ALICE_JWT");
    }

    private void loginAsBob() {
        otherUser = "Alice";

        client.login("BOB_JWT");
    }

    @SuppressLint("MissingPermission")
    private void startCall() {
        client.call(otherUser, NexmoCallHandler.SERVER, new NexmoRequestListener<NexmoCall>() {
            @Override
            public void onError(@NonNull NexmoApiError nexmoApiError) {

            }

            @Override
            public void onSuccess(@Nullable NexmoCall call) {
                runOnUiThread(() -> {
                    hideUI();
                    endCallButton.setVisibility(View.VISIBLE);
                    waitingForIncomingCallTextView.setVisibility(View.VISIBLE);
                });

                onGoingCall = call;

                onGoingCall.addCallEventListener(new NexmoCallEventListener() {
                    @Override
                    public void onMemberStatusUpdated(NexmoCallMemberStatus callStatus, NexmoMember NexmoMember) {
                        if (callStatus == NexmoCallMemberStatus.COMPLETED || callStatus == NexmoCallMemberStatus.CANCELLED) {
                            onGoingCall = null;

                            runOnUiThread(() -> {
                                    hideUI();
                                    startCallButton.setVisibility(View.VISIBLE);
                                }
                            );
                        }
                    }

                    @Override
                    public void onMuteChanged(NexmoMediaActionState nexmoMediaActionState, NexmoMember NexmoMember) {

                    }

                    @Override
                    public void onEarmuffChanged(NexmoMediaActionState nexmoMediaActionState, NexmoMember NexmoMember) {

                    }

                    @Override
                    public void onDTMF(String s, NexmoMember NexmoMember) {

                    }
                });
            }
        });
    }

    @SuppressLint("MissingPermission")
    private void answerCall() {
        onGoingCall.answer(new NexmoRequestListener<NexmoCall>() {
            @Override
            public void onError(@NonNull NexmoApiError nexmoApiError) {

            }

            @Override
            public void onSuccess(@Nullable NexmoCall nexmoCall) {
                runOnUiThread(() -> {
                    hideUI();
                    endCallButton.setVisibility(View.VISIBLE);
                });
            }
        });
    }

    private void rejectCall() {
        onGoingCall.hangup(new NexmoRequestListener<NexmoCall>() {
            @Override
            public void onError(@NonNull NexmoApiError nexmoApiError) {

            }

            @Override
            public void onSuccess(@Nullable NexmoCall nexmoCall) {
                runOnUiThread(() -> {
                    hideUI();
                    startCallButton.setVisibility(View.VISIBLE);
                    waitingForIncomingCallTextView.setVisibility(View.VISIBLE);
                });
            }
        });
        onGoingCall = null;
    }

    private void endCall() {
        onGoingCall.hangup(new NexmoRequestListener<NexmoCall>() {
            @Override
            public void onError(@NonNull NexmoApiError nexmoApiError) {

            }

            @Override
            public void onSuccess(@Nullable NexmoCall nexmoCall) {
                runOnUiThread(() -> {
                    hideUI();
                    startCallButton.setVisibility(View.VISIBLE);
                    waitingForIncomingCallTextView.setVisibility(View.VISIBLE);
                });
            }
        });

        onGoingCall = null;
    }
}

File: app-to-app-kotlin/app-to-app/app/src/main/java/com/vonage/tutorial/apptoapp/MainActivity.kt
package com.vonage.tutorial.apptoapp

import android.Manifest
import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.view.forEach
import com.vonage.voice.api.*

class MainActivity : AppCompatActivity() {

    private val aliceJWT = ""
    private val bobJWT = ""
    private lateinit var client: VoiceClient
    private var otherUser: String = ""
    private var onGoingCall: VoiceCall? = null
    private var callInvite: VoiceInvite? = null

    private lateinit var connectionStatusTextView: TextView
    private lateinit var waitingForIncomingCallTextView: TextView
    private lateinit var loginAsAlice: Button
    private lateinit var loginAsBob: Button
    private lateinit var startCallButton: Button
    private lateinit var answerCallButton: Button
    private lateinit var rejectCallButton: Button
    private lateinit var endCallButton: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // request permissions
        val callsPermissions = arrayOf(Manifest.permission.RECORD_AUDIO)
        ActivityCompat.requestPermissions(this, callsPermissions, 123)

        // init views
        connectionStatusTextView = findViewById(R.id.connectionStatusTextView)
        waitingForIncomingCallTextView = findViewById(R.id.waitingForIncomingCallTextView)
        loginAsAlice = findViewById(R.id.loginAsAlice)
        loginAsBob = findViewById(R.id.loginAsBob)
        startCallButton = findViewById(R.id.startCallButton)
        answerCallButton = findViewById(R.id.answerCallButton)
        rejectCallButton = findViewById(R.id.rejectCallButton)
        endCallButton = findViewById(R.id.endCallButton)

        loginAsAlice.setOnClickListener { loginAsAlice() }
        loginAsBob.setOnClickListener { loginAsBob() }
        answerCallButton.setOnClickListener { answerCall() }
        rejectCallButton.setOnClickListener { rejectCall() }
        endCallButton.setOnClickListener { endCall() }
        startCallButton.setOnClickListener { startCall() }

        client = VoiceClient(this.application.applicationContext)
        client.setConfig(ClientConfig(ConfigRegion.US))

        client.setCallInviteListener { invite ->
            callInvite = invite
            runOnUiThread {
                hideUI()
                answerCallButton.visibility = View.VISIBLE
                rejectCallButton.visibility = View.VISIBLE
            }
        }
    }

    private fun hideUI() {
        val content = findViewById<LinearLayout>(R.id.content)
        content.forEach { it.visibility = View.GONE }
    }

    private fun loginAsAlice() {
        otherUser = "Bob"
        client.createSession(aliceJWT, null) {
                err, sessionId ->
            when {
                err != null -> {
                    hideUI()
                    connectionStatusTextView.visibility = View.VISIBLE
                    connectionStatusTextView.text = err.localizedMessage
                }
                else -> {
                    hideUI()
                    connectionStatusTextView.visibility = View.VISIBLE
                    connectionStatusTextView.text = "Connected"
                    startCallButton.visibility = View.VISIBLE
                    waitingForIncomingCallTextView.visibility = View.VISIBLE
                }
            }
        }
    }

    private fun loginAsBob() {
        otherUser = "Alice"
        client.createSession(bobJWT, null) {
                err, sessionId ->
            when {
                err != null -> {
                    hideUI()
                    connectionStatusTextView.visibility = View.VISIBLE
                    connectionStatusTextView.text = err.localizedMessage
                }
                else -> {
                    hideUI()
                    connectionStatusTextView.visibility = View.VISIBLE
                    connectionStatusTextView.text = "Connected"
                    startCallButton.visibility = View.VISIBLE
                    waitingForIncomingCallTextView.visibility = View.VISIBLE
                }
            }
        }
    }

    @SuppressLint("MissingPermission")
    fun startCall() {
        client.serverCall(mapOf("callee" to otherUser)) {
                err, outboundCall ->
            when {
                err != null -> {
                    connectionStatusTextView.text = err.localizedMessage
                }
                else -> {
                    onGoingCall = outboundCall
                    hideUI()
                    endCallButton.visibility = View.VISIBLE
                }
            }
        }
    }


    @SuppressLint("MissingPermission")
    private fun answerCall() {
        callInvite?.answer {
            err, incomingCall ->
            when {
                err != null -> {
                    connectionStatusTextView.text = err.localizedMessage
                }
                else -> {
                    onGoingCall = incomingCall
                    hideUI()
                    endCallButton.visibility = View.VISIBLE
                }
            }
        }
    }

    private fun rejectCall() {
        callInvite?.reject {
                err ->
            when {
                err != null -> {
                    connectionStatusTextView.text = err.localizedMessage
                }
                else -> {
                    hideUI()
                    startCallButton.visibility = View.VISIBLE
                    waitingForIncomingCallTextView.visibility = View.VISIBLE
                }
            }
        }
        onGoingCall = null
    }

    private fun endCall() {
        onGoingCall?.hangup {
                err ->
            when {
                err != null -> {
                    connectionStatusTextView.text = err.localizedMessage
                }
                else -> {
                    hideUI()
                    startCallButton.visibility = View.VISIBLE
                    waitingForIncomingCallTextView.visibility = View.VISIBLE
                }
            }
        }
        onGoingCall = null
    }
}
